README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# covidphtext: Utilities to Extract Text Data From COVID-19-Related Resolutions and Policies From the Philippines <img src="man/figures/covidphtext.png" width="200px" align="right" />

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN
status](https://www.r-pkg.org/badges/version/covidphtext)](https://CRAN.R-project.org/package=covidphtext)
[![CRAN](https://img.shields.io/cran/l/covidphtext.svg)](https://CRAN.R-project.org/package=covidphtext)
[![CRAN](http://cranlogs.r-pkg.org/badges/covidphtext)](https://CRAN.R-project.org/package=covidphtext)
[![CRAN](http://cranlogs.r-pkg.org/badges/grand-total/covidphtext)](https://CRAN.R-project.org/package=covidphtext)
[![R build
status](https://github.com/como-ph/covidphtext/workflows/R-CMD-check/badge.svg)](https://github.com/como-ph/covidphtext/actions)
[![test-coverage](https://github.com/como-ph/covidphtext/workflows/test-coverage/badge.svg)](https://github.com/como-ph/covidphtext/actions?query=workflow%3Atest-coverage)
[![Codecov test
coverage](https://codecov.io/gh/como-ph/covidphtext/branch/master/graph/badge.svg)](https://codecov.io/gh/como-ph/covidphtext?branch=master)
[![CodeFactor](https://www.codefactor.io/repository/github/como-ph/covidphtext/badge)](https://www.codefactor.io/repository/github/como-ph/covidphtext)
[![DOI](https://zenodo.org/badge/265376181.svg)](https://zenodo.org/badge/latestdoi/265376181)
<!-- badges: end -->

To assess possible impact of various COVID-19 prediction models on
Philippine government response, text from various resolutions issued by
the Inter-agency Task Force for the Management of Emerging Infectious
Diseases (IATF) has been collected using data mining approaches
implemented in R. This package includes functions used for this data
mining process and datasets of text that have been collected and
processed for use in text analysis.

## Installation

`covidphtext` is not yet available on CRAN. It is currently in active
development stage. Installation of `covidphtext` at this point is only
possible through its development version via
[GitHub](https://github.com/como-ph/covidphtext):

``` r
if(!require(remotes)) install.packages("remotes")
remotes::install_github("como-ph/covidphtext")
```

then load the package:

``` r
library(covidphtext)
```

## Usage

### Datasets

`covidphtext` currently has `69` datasets of which `65` are
COVID-19-related resolutions and policies in the Philippines made by the
Inter-Agency Task Force for the Management of Emerging Infectious
Diseases (IATF), 1 is the Omnibus Guidelines on the Implementation of
Community Quarantine in the Philippines released by the IATF and 2 are
reference lists of links to these resolutions and guidelines.

A description of the available datasets can be found
[here](https://como-ph.github.io/covidphtext/reference/index.html#section-datasets).

The IATF resolutions are officially available from two online sources:
1) The Department of Health (DoH)
[website](http://www.doh.gov.ph/COVID-19/IATF-Resolutions/); and, 2) The
Philippines Official Gazette
[website](https://www.officialgazette.gov.ph/section/laws/other-issuances/inter-agency-task-force-for-the-management-of-emerging-infectious-diseases-resolutions/).
The DOH website currently only holds IATF resolutions starting from
resolution number 9 and later. The Official Gazette on the other hand
contains resolutions 1 to the most current.

To get a list of the IATF resolutions that are available from the DOH
website, the function `get_iatf_links()` can be used as follows:

``` r
get_iatf_links()
#> # A tibble: 74 x 7
#>       id title              date       source type  url               checked   
#>    <dbl> <chr>              <date>     <chr>  <chr> <chr>             <date>    
#>  1     9 Recommendations f… 2020-03-03 IATF   reso… https://doh.gov.… 2020-08-04
#>  2    10 Recommendations f… 2020-03-09 IATF   reso… https://doh.gov.… 2020-08-04
#>  3    11 Recommendations f… 2020-03-12 IATF   reso… https://doh.gov.… 2020-08-04
#>  4    12 Recommendations f… 2020-03-13 IATF   reso… https://doh.gov.… 2020-08-04
#>  5    13 Recommendations f… 2020-03-17 IATF   reso… https://doh.gov.… 2020-08-04
#>  6    14 Resolutions Relat… 2020-03-20 IATF   reso… https://doh.gov.… 2020-08-04
#>  7    15 Resolutions Relat… 2020-03-25 IATF   reso… https://doh.gov.… 2020-08-04
#>  8    16 Additional Guidel… 2020-03-30 IATF   reso… https://doh.gov.… 2020-08-04
#>  9    17 Recommendations R… 2020-03-30 IATF   reso… https://doh.gov.… 2020-08-04
#> 10    18 Recommendations R… 2020-04-01 IATF   reso… https://doh.gov.… 2020-08-04
#> # … with 64 more rows
```

Given that the DOH website doesn’t have the first 8 resolutions, this
function will soon be deprecated in favour of the newer function below
that interfaces with the Official Gazette.

A table of all the IATF resolutions and the URLs to download them can be
generated using the newer function `get_iatf_gazette()` as follows:

``` r
list_iatf_pages(base = "https://www.officialgazette.gov.ph/section/laws/other-issuances/inter-agency-task-force-for-the-management-of-emerging-infectious-diseases-resolutions/", 
                pages = 1:6) %>%
  get_iatf_pages() %>%
  get_iatf_gazette()
```

    #> # A tibble: 73 x 7
    #>       id title              date       source type  url               checked   
    #>    <dbl> <chr>              <date>     <chr>  <chr> <chr>             <date>    
    #>  1    60 RESOLUTION NO. 60… 2020-07-30 IATF   reso… https://www.offi… 2020-08-02
    #>  2    60 RESOLUTION NO. 60… 2020-07-30 IATF   reso… https://www.offi… 2020-08-02
    #>  3    59 RESOLUTION NO. 59… 2020-07-28 IATF   reso… https://www.offi… 2020-08-02
    #>  4    58 RESOLUTION NO. 58… 2020-07-23 IATF   reso… https://www.offi… 2020-08-02
    #>  5    57 RESOLUTION NO. 57… 2020-07-21 IATF   reso… https://www.offi… 2020-08-02
    #>  6    NA OMNIBUS GUIDELINE… 2020-07-16 IATF   reso… https://www.offi… 2020-08-02
    #>  7    56 RESOLUTION NO. 56… 2020-07-16 IATF   reso… https://www.offi… 2020-08-02
    #>  8    55 RESOLUTION NO. 55… 2020-07-14 IATF   reso… https://www.offi… 2020-08-02
    #>  9    55 RESOLUTION NO. 55… 2020-07-14 IATF   reso… https://www.offi… 2020-08-02
    #> 10    54 RESOLUTION NO. 54… 2020-07-11 IATF   reso… https://www.offi… 2020-08-02
    #> # … with 63 more rows

The actual PDF of the IATF resolutions/s can be downloaded using the
`get_iatf_pdfs()` function. For example, to download IATF Resolution
No. 29, the following command is issued:

``` r
get_iatf_pdfs(links = iatfLinks, id = 29)
#>                                                                iatfResolution29 
#> "/var/folders/fk/s0yv8hhn2cs_nfsmzhm4dmhc0000gn/T//RtmpOtmTrk/file263d25fd8821"
```

The command downloads the PDF of the specified IATF Resolution into a
temporary directory (using `tempdir()` function). The output of the
`get_iatf_pdfs()` function is a named character vector of directory
path/s to downloaded PDFs as shown above. The names of the character
vector correspond to the resolution number. These paths can then be used
when working with these files.

The `get_iatf_pdfs()` function interfaces with both the DOH and The
Official Gazette website.

### Concatenating text datasets

The datasets described above can be processed and analysed on their own
or as a combined corpus of text data. `covidphtext` provides convenience
functions that concatenates all or specific text datasets available from
the `covidphtext` package.

#### Concatenating datasets based on a specific search term

The `combine_docs` function allows the user to specify search terms to
use in identifying datasets provided by the `covidphtext` package. The
`docs` argument allows the specification of a vector of search terms to
use to identify the names of datasets to concatenate. If the name/s of
the datasets contain these search terms, the datasets with these name/s
will be returned.

``` r
combine_docs(docs = "resolution")
#> # A tibble: 7,728 x 7
#>    linenumber text                       source type       id section date      
#>         <int> <chr>                      <chr>  <chr>   <dbl> <chr>   <date>    
#>  1          1 Republic of the Philippin… IATF   resolu…     1 heading 2020-01-28
#>  2          2 Department of Health       IATF   resolu…     1 heading 2020-01-28
#>  3          3 Office of the Secretary    IATF   resolu…     1 heading 2020-01-28
#>  4          4 Inter-Agency Task Force f… IATF   resolu…     1 heading 2020-01-28
#>  5          5 Emerging Infectious Disea… IATF   resolu…     1 heading 2020-01-28
#>  6          6 28 January 2020            IATF   resolu…     1 heading 2020-01-28
#>  7          7 Resolution No. 01          IATF   resolu…     1 heading 2020-01-28
#>  8          8 Series of 2020             IATF   resolu…     1 heading 2020-01-28
#>  9          9 Recommendations for the M… IATF   resolu…     1 heading 2020-01-28
#> 10         10 Novel Coronavirus Situati… IATF   resolu…     1 heading 2020-01-28
#> # … with 7,718 more rows
```

The `combine_iatf` function is a specialised wrapper of the
`combine_docs` function that specifically returns datasets containing
IATF resolutions. An additional argument `res` allows users to specify
which IATF resolutions to return. To get IATF resolution 10, 11, and 12,
the following call to `combine_iatf` is made as follows:

``` r
combine_iatf(docs = "resolution", res = 10:12)
#> # A tibble: 351 x 7
#>    linenumber text                        source type      id section date      
#>         <int> <chr>                       <chr>  <chr>  <dbl> <chr>   <date>    
#>  1          1 Republic of the Philippines IATF   resol…    10 heading 2020-03-09
#>  2          2 DOH DFA DILG DOJ DOLE DOT … IATF   resol…    10 heading 2020-03-09
#>  3          3 Inter-Agency Task Force fo… IATF   resol…    10 heading 2020-03-09
#>  4          4 Emerging Infectious Diseas… IATF   resol…    10 heading 2020-03-09
#>  5          5 Resolution No. 10           IATF   resol…    10 heading 2020-03-09
#>  6          6 Series of 2020              IATF   resol…    10 heading 2020-03-09
#>  7          7 9 March 2020                IATF   resol…    10 heading 2020-03-09
#>  8          8 Recommendations for the ma… IATF   resol…    10 heading 2020-03-09
#>  9          9 disease 2019 (COVID-19) si… IATF   resol…    10 heading 2020-03-09
#> 10         10 WHEREAS, Section 15 of Art… IATF   resol…    10 preamb… 2020-03-09
#> # … with 341 more rows
```

To check if only resolutions 10 to 12 have been returned:

``` r
combine_iatf(docs = "resolution", res = 10:12)[ , c("type", "id")]
#> # A tibble: 351 x 2
#>    type          id
#>    <chr>      <dbl>
#>  1 resolution    10
#>  2 resolution    10
#>  3 resolution    10
#>  4 resolution    10
#>  5 resolution    10
#>  6 resolution    10
#>  7 resolution    10
#>  8 resolution    10
#>  9 resolution    10
#> 10 resolution    10
#> # … with 341 more rows
```


NEWS.md
# covidphtext v0.2.1

* Re-assigned linenumbers to concatenated datasets when using `combine_` functions

* Re-worked some get functions to fail gracefully when connection with webpage fails

* Added `get_iatf_pdfs` and `get_iatf_page` functions

* Added new resolutions up to 1 August 2020

* Added `codefactor` check

# covidphtext v0.2.0

* Updated `get_iatf_pdf` syntax such that is now takes on as first argument a links table produced when making a call to either `get_iatf_links` or `get_iatf_gazette` allowing for piped operations between these set of functions

* Updated IATF resolutions datasets structure to include sections of resolutions

* Added missing resolutions no. 30 and 30A

# covidphtext v0.1.0

* Created functions to extract download information for the various IATF resolutions issued either through the Department of Health or the Official Gazette of the Republic of the Philippines

* Added text datasets for each available IATF resolution (as of 16 June 2020, there were 46 resolutions)

* Added CI/CD configuration using GitHub Actions [![R build status](https://github.com/como-ph/covidphtext/workflows/R-CMD-check/badge.svg)](https://github.com/como-ph/covidphtext/actions)

* Added coverage tests ![test-coverage](https://github.com/como-ph/covidphtext/workflows/test-coverage/badge.svg) [![codecov](https://codecov.io/gh/como-ph/covidphtext/branch/master/graph/badge.svg)](https://codecov.io/gh/como-ph/covidphtext)

* Added a package website using `pkgdown`


R/01-get_iatf.R
################################################################################
#
#'
#' Get list of links to IATF resolutions from Department of Health website
#'
#' @param base URL to the IATF resolutions webpage. This is currently at
#'   \url{https://www.doh.gov.ph/COVID-19/IATF-Resolutions}
#'
#' @return A tibble containing absolute links to all the current IATF
#'   resolutions at time of extraction. The tibble contains the following
#'   information:
#'
#' \itemize{
#'   \item \code{id}{Resolution number}
#'   \item \code{title}{Title of resolution}
#'   \item \code{date}{Date (in <YYYY-MM-DD> format) resolution was issued}
#'   \item \code{source}{Source of resolution. This is by default from IATF}
#'   \item \code{type}{Type of document. This is by default a resolution}
#'   \item \code{url}{Absolute URL for PDF of resolution}
#'   \item \code{checked}{Date (in <YYYY-MM-DD format) table was extracted. This
#'     is by default provided by \code{Sys.Date()}}
#' }
#'
#' @examples
#' get_iatf_links(base = "http://www.doh.gov.ph/COVID-19/IATF-Resolutions")
#'
#' @export
#'
#
################################################################################

get_iatf_links <- function(base = "https://www.doh.gov.ph/COVID-19/IATF-Resolutions") {
  ## Get page table
  xx <- xml2::read_html(base) %>%
    rvest::html_nodes(css = ".panel .view-content .views-table") %>%
    rvest::html_table()

  ## Convert Resolution No. to numeric
  xx[[1]]$`Resolution No.` <- as.numeric(xx[[1]]$`Resolution No.`)

  ## Convert date to date class with format YYYY-MM-DD
  xx[[1]]$Date <- lubridate::mdy(xx[[1]]$Date)

  ## Get href links per resolution
  yy <- xml2::read_html(base) %>%
    rvest::html_nodes(css = ".panel .view-content .views-field a") %>%
    rvest::html_attr(name = "href")

  ## Extract Resolution number from links
  id <- stringr::str_remove_all(string = yy, pattern = "%20")
  id <- stringr::str_extract(string = id, pattern = "[0-9]+")

  ## Add links id to links
  yy <- data.frame(id,
                   source = "IATF",
                   type = "resolution",
                   url = yy,
                   checked = Sys.Date(),
                   stringsAsFactors = FALSE)

  ## Merge links with page table
  linkTable <- merge(xx[[1]], yy, by.x = "Resolution No.", by.y = "id")

  ## Rename fields in linkTable
  names(linkTable) <- c("id", "title", "date", "source", "type", "url", "checked")

  ## Convert to tibble
  linkTable <- tibble::tibble(linkTable)

  ## Return
  return(linkTable)
}



################################################################################
#
#'
#' Get a specific IATF resolution PDF file
#'
#' @param links A links/URLs tibble produced by either \code{get_iatf_links()}
#'   or \code{get_iatf_gazette()} functions
#' @param id A vector of number/s of IATF resolution/s to be retrieved
#'
#' @return A temporary directory path or a vector of temporary directory paths
#'   pointing to a temporary file/s for PDF of IATF resolution/s required
#'
#' @examples
#' get_iatf_pdf(links = iatfLinks, id = 29)
#'
#' @export
#'
#
################################################################################

get_iatf_pdf <- function(links, id) {
  ## Check that id is numeric
  if(class(id) == "character" | class(id) == "factor") {
    stop("Numeric value required for id. Please try again.", .call = TRUE)
  }

  ## Get href links per resolution
  href <- links$url

  ## Get available resolutions by resolution ID
  availableIDs <- links$id

  idText <- paste(availableIDs, collapse = ", ")

  ## Get availableIDs specified by id
  linksID <- availableIDs[which(availableIDs %in% id)]

  if(length(linksID) == 0) {
    stop(paste("Value/s for id is/are outside of what is currently available. Only resolutions ",
               idText, " are available. Please try again", sep = ""), call. = TRUE)
  }

  if(length(linksID) < length(id)) {
    warning(paste("Not all values for id correspond to an available resolution. Returning only ",
                   length(linksID), " of ", length(id), " resolutions requested."), call. = TRUE)
  }

  ## Concatenating vector for directory paths
  paths <- NULL

  ## Cycle through identifiers
  for(i in linksID) {

    ## Create temporary destfile
    destfile <- paste("iatfResolution", i, ".pdf", sep = "")

    ## Get URL for current resolution id
    link <- href[which(availableIDs == i)]

    ## Download resolution with current id
    curl::curl_download(url = link,
                        destfile = paste(tempdir(), destfile, sep = "/"))
  }

  ## Create vector of directory paths of temporary files
  paths <- paste(tempdir(), "/iatfResolution", linksID, ".pdf", sep = "")

  ## Return paths
  return(paths)
}


################################################################################
#
#'
#' Get list of links to IATF resolutions from Philippines Official Gazette
#'
#' @param base URL to the IATF resolutions webpage in the Official Gazette.
#'   This is currently at \url{https://www.officialgazette.gov.ph/section/laws/other-issuances/inter-agency-task-force-for-the-management-of-emerging-infectious-diseases-resolutions/}
#' @param pages A vector of paginated webpages in which the IATF resolutions
#'   are available. This currently defaults to 1:6 as there are currently 6
#'   pages.
#'
#' @return A tibble containing absolute links to all the current IATF
#'   resolutions at time of extraction from the Official Gazette. The tibble
#'   contains the following information:
#'
#' \itemize{
#'   \item \code{id}{Resolution number}
#'   \item \code{title}{Title of resolution}
#'   \item \code{date}{Date (in <YYYY-MM-DD> format) resolution was issued}
#'   \item \code{source}{Source of resolution. This is by default from IATF}
#'   \item \code{type}{Type of document. This is by default a resolution}
#'   \item \code{url}{Absolute URL for PDF of resolution}
#'   \item \code{checked}{Date (in <YYYY-MM-DD format) table was extracted. This
#'     is by default provided by \code{Sys.Date()}}
#' }
#'
#' @examples
#' base <- "https://www.officialgazette.gov.ph/section/laws/other-issuances"
#' agency <- "inter-agency-task-force-for-the-management-of-emerging-infectious-diseases-resolutions/"
#'
#' get_iatf_gazette(base = paste(base, agency, sep = "/"), pages = 1)
#'
#' @export
#'
#
################################################################################

get_iatf_gazette <- function(base, pages = 1:5) {
  ## Create concatenating objects
  iatfID <- NULL
  iatfTitle <- NULL
  iatfDate <- NULL
  iatfURL <- NULL

  ## Cycle through resolution pages
  for(i in pages) {
    ## Create current URL page
    urlPage <- xml2::read_html(x = paste(base, "page", i,  sep = "/"))

    ## Get current page resolutions ID
    iatfID <- c(iatfID,
                urlPage %>%
                  rvest::html_nodes(css = ".large-8 .entry-title a") %>%
                  rvest::html_text() %>%
                  stringr::str_remove(pattern = "2020") %>%
                  stringr::str_extract(pattern = "[0-9]+")) %>%
                  as.numeric()

    ## Get current page resolutions Title
    iatfTitle <- c(iatfTitle,
                   urlPage %>%
                     rvest::html_nodes(css = ".large-8 p") %>%
                     rvest::html_text())

    ## Get current page resolutions Date
    iatfDate <- c(iatfDate,
                  urlPage %>%
                    rvest::html_nodes(css = ".large-8 .published") %>%
                    rvest::html_text())

    ## Get current page resolutions URL
    iatfURL <- c(iatfURL,
                 urlPage %>%
                   rvest::html_nodes(css = ".large-8 .entry-title a") %>%
                   rvest::html_attr(name = "href"))
  }

  iatfURLs <- NULL

  for(i in iatfURL) {
    resolutionPage <- xml2::read_html(x = i)

    pdfLink <- resolutionPage %>%
      rvest::html_nodes(css = "#resource a") %>%
      rvest::html_attr(name = "href")

    pdfLink <- pdfLink[1]

    iatfURLs <- c(iatfURLs, pdfLink)
  }

  ## Concatenate IATF list
  iatfLinksGazette <- data.frame(id = ifelse(stringr::str_detect(string = iatfTitle,
                                                                 pattern = "OMNIBUS"),
                                             NA, iatfID),
                                 title = iatfTitle,
                                 date = lubridate::mdy(iatfDate),
                                 source = "IATF",
                                 type = "resolution",
                                 url = iatfURLs,
                                 checked = Sys.Date(),
                                 stringsAsFactors = FALSE)

  ## Convert to tibble
  iatfLinksGazette <- tibble::tibble(iatfLinksGazette)

  ## Return IATF list
  return(iatfLinksGazette)
}
